ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala

```diff
@@ -101,8 +103,40 @@ object Conversions {
         .fold(msg => throw Err.InvalidSubmission(msg), identity)
     )
 
-  def decodeIdentifier(protoIdent: ValueOuterClass.Identifier): Ref.Identifier =
-    assertDecode("Identifier", ValueCoder.decodeIdentifier(protoIdent))
+  def encodeIdentifier(id: Ref.Identifier): Identifier =
+    Identifier
+      .newBuilder()
+      .setPackageId(id.packageId)
+      .addAllModuleName((id.qualifiedName.module.segments.toSeq: Seq[String]).asJava)
+      .addAllName((id.qualifiedName.name.segments.toSeq: Seq[String]).asJava)
+      .build()
+
+  @throws[Err.DecodeError]
+  def decodeIdentifier(id: Identifier): Ref.Identifier = {
+    val errorOrRefId: Either[Err.DecodeError, Ref.Identifier] = for {
+      pkgId <- PackageId
+        .fromString(id.getPackageId)
+        .left
+        .map(_ => Err.DecodeError("Identifier", s"Invalid package ID: ${id.getPackageId}"))
+
+      moduleSegments = id.getModuleNameList.asScala
+      module <- ModuleName
+        .fromSegments(id.getModuleNameList.asScala)
+        .left
+        .map(_ => Err.DecodeError("Identifier", s"Invalid module segments: $moduleSegments"))
+
+      nameSegments = id.getNameList.asScala
+      name <- DottedName
+        .fromSegments(nameSegments)
+        .left
+        .map(_ => Err.DecodeError("Identifier", s"Invalid name segments: $nameSegments"))
+    } yield Ref.Identifier(pkgId, QualifiedName(module, name))
+
+    errorOrRefId match {
+      case Right(refId) => refId
+      case Left(decodeError) => throw decodeError
+    }
```

Comment:
Any specific reason why we get rid of the Either here and not just return the Either?

Reply:
This is the convention in `kvutils`/`Conversions` but there is just one usage here, so I have no strong opinion on this.

Reply:
Actually, let me change this.